### ISSUE TITLE
Context data for pywbem exceptions

### ISSUE THREAD

**Issue by andy-maier** [pywbem/pywbem#1423]
Recent runs in the SNIA center showed that the information we have in exceptions that happen, is not optimal, particularly when the server is returning generic error messages.

The idea would be to add context and/or detail data to the pywbem exceptions that are raised.

For example:
- ParseError should have the CIM-XML reply
- CIMError should have the operation that was issued.

This list needs to be completed.

**Comment by andy-maier** [comment]
A more concrete proposal is the following, looking at each pywbem exception that is actually raised:

The following exceptions are only raised in cim_http.py:
* ConnectionError: No need to add further context
* AuthError: No need to add further context
* TimeoutError: No need to add further context
* HTTPError: Add request XML

The following exceptions are only raised in _listener.py:
* VersionError: Add request XML

The following exceptions are raised only after receiving a CIM-XML response:
* CIMXMLParseError: Add request XML and response XML
* XMLParseError: Add request XML and response XML

CIMError is raised in multiple situations:
* In cim_operations.py, when a CIM operation fails: Add request XML
* In mof_compiler.py: It already adds file_line as inofficial attribute: Add filename, lineno, MOF line
* In _subscription_manager.py, when filter or listener are removed: No need to add further context
* In _server.py: TBD
* In mock support: TBD

The following are raised in _server.py and _valuemapping.py:
* ModelError: No need to add further context

**Comment by andy-maier** [comment]
Karl volunteers to specify the proposal for what to change in CIMError when raised in _server.py and in the mock support.

**Comment by KSchopmeyer** [comment]
_server.py  CIMError is generated by this pywbem module in the following cases:

1. Line 452, in create_namespace, method 1 0 or multiple instances of the requested profile when exactly one should be returned.

2. Line 479 in create_namespace(...) if the invokeMethod to create the namespace returns with return value other than zero

3. line 592 in delete_namespace(...) if the defined namespace is not in self.namespaces which means that the server did not return in in the list of namespaces.

4. Line 613 in delete_namespace either there are classes or qualifiers in the namespace to be deleted.

5. Line 1108,  in _determine_interop_ns, The method cannot determine an interop namespace.

6. Line 1243, in _determine_brand when more or less than one CIM_ObjectManager instances are received.

In all other cases it is driven by the receipt of a corresponsing CIMError from the server which we simply pass on.

We have an inherent conflict here in that:

a. Generally we propose that we expect Error or CIMError from the server so that the client need protect is calls only with these exceptions so if we were to modify to us a pywbem exception it would mean changes for the user.

b. Since the server is not generating these exceptions, the user may be confused by the reception of CIMError exceptions.

I propose that the simplest solution is:

a. Move the responsibility for the issues in delete_instance to the server.  We just issue the delete and the server is responsible if it shoots itself. Therefore we would not test for classes and qualifiers at all or follow the proposal below for this exception if that seriously worries us.   I actually prefer that we apply b to this exception also.

b. Add some test in the others to indicate that they pywbem generated and let it go at that.  For example:
 "Namespace class could not be determined" becomes "pywbem client could not determine Namespace class"

"Unexpected number of CIM_ObjectManager instances: " becomes "pywbem client received unexpected number of CIM_ObjectManager instances:"

etc.

**Comment by KSchopmeyer** [comment]
Mock support is, in effect a server.  As part of the refactor to separate out the store, repository, mocker into separate components I will review to assure that these to match the rules in DMTF spec. Other than that, I think that they are correct in that they should be mimicking what a server should return.

**Comment by andy-maier** [comment]
Andy to digest Karl's proposal

**Comment by andy-maier** [comment]
Since this issue is about adding context to the exceptions, I suggest to move changes for which exceptions are raised into a separate issue.

**Comment by andy-maier** [comment]
Here are a couple of changes relative to the initial assessment:
* PR #1995 (1.0.0) already removed the use of CIMError for reporting MOF compiler errors.
* The only remaining raises of CIMError in mof_compiler.py are in MOFWBEMConnection, where they will be handled as part of refactoring the mock support and repository interfaces (e.g. PR #2080).
* PR #1573 (0.13.0) already added request and response to the ParseError exception and its subclasses CIMXMLParseError and XMLParseError.
* VersionError is only raised in _listener.py to report a protocol or DTD version mismatch in the received export message request. That exception is caught again by the listener code to handle the mismatch by returning corresponding error information to the server in the export message response. That means adding CIM-XML request data to the VersionError exception is not worthwhile at this point.
* HTTPError is raised in a situation where pywbem detects an invalid content-type header. For that case, it helps storing the response data in addition to the request data.
* On Karl's proposal, above: In _server.py, a number of raises of CIMError should actually be raises of ModelError. The remaining two raises of CIMError are in WBEMServer.delete_namespace() for namespace not found and namespace not empty. We could leave that to the server as Karl suggests, but (1) I'm hesitant on that and (2) it should be in a eparate PR

That results in the following changes to be done:

* Streamline the way ParseError stored request XML and response XML with the new, more general approach. (DONE)
* HTTPError:
  - Add request XML and response XML to the exception class (DONE)
  - When raised in cim_http.py: Always store request XML, and store response XML when detecting an invalid content-type header (DONE)
* CIMError:
  - Add request XML to the exception class (DONE)
  - When raised in cim_operations.py when a CIM operation fails: Store request XML (DONE)
  - When raised in _server.py: TBD (TODO)
  - When raised in mock support: TBD (TODO)

**Comment by andy-maier** [comment]
The changes mentioned in the previous comment have all been done in PR TBD

**Comment by andy-maier** [comment]
We agreed to the proposed implementation, except for the "misuse" of HTTPError for reporting invalid content-type, where we probably should use a new HeaderParseError subclass of ParseError
